### DISCUSSION TITLE
JTFS of A.M. cosine [Case Study]

### DISCUSSION THREAD

**Discussion by OverLordGoldDragon** [kymatio/kymatio#710] [Show and tell]
Below are visuals for JTFS of a cosine amplitude-modulated cosine, along with the code. 

<img src="https://user-images.githubusercontent.com/16495490/113690319-518b7e80-96dc-11eb-981f-a10982aeb1a4.png" width="800">

## `average=False`, no subsampling

<img src="https://user-images.githubusercontent.com/16495490/113827688-a5599e80-9794-11eb-9dc3-5c7910d6c31e.gif" width="600">
<img src="https://user-images.githubusercontent.com/16495490/113827694-a68acb80-9794-11eb-80cb-8e365e464667.gif" width="380">
<img src="https://user-images.githubusercontent.com/16495490/113827695-a68acb80-9794-11eb-85be-a1b01ab39720.gif" width="380">

## `average=True`, no subsampling

<img src="https://user-images.githubusercontent.com/16495490/113819259-1fd0f100-978a-11eb-9e19-f0b3c01d3f72.gif" width="600">
<img src="https://user-images.githubusercontent.com/16495490/113819266-21021e00-978a-11eb-8d0e-4f831f9c4d0e.gif" width="380">
<img src="https://user-images.githubusercontent.com/16495490/113819267-21021e00-978a-11eb-8183-e95e6bddb981.gif" width="380">

## `average=True`, max subsampling

<img src="https://user-images.githubusercontent.com/16495490/113707330-cf598500-96f0-11eb-830d-80b3e208b5ff.gif" width="600">
<img src="https://user-images.githubusercontent.com/16495490/113706504-cf0cba00-96ef-11eb-8fda-3048f21763a3.gif" width="380">
<img src="https://user-images.githubusercontent.com/16495490/113706555-de8c0300-96ef-11eb-9dc2-8341467a2d57.gif" width="380">

## Coefficients explained

JTFS convolves 2D wavelets over the modulus of the Continuous Wavelet Transform. The wavelets are defined independently along frequency and time, permitting separable convolution - but also _separable interpretation_, which is given here.

### 1. First visualize first-order coefficients

![image](https://user-images.githubusercontent.com/16495490/113699128-3cb3e880-96e6-11eb-9527-e86f217fb183.png)

As expected, `abs(CWT)` captures the A.M. envelope.

### 2. The next step is second-order time scattering

Here we convolve each second order wavelet with _every row of `abs(CWT)`_. Where do we expect to see maximum second-order coefficients, and what will they look like? Reminder of second-order wavelets:

![image](https://user-images.githubusercontent.com/16495490/113697350-1f7e1a80-96e4-11eb-8c9f-46656d3ca451.png)

with different colors corresponding to different (increasing) subsampling rates. The wavelets are ordered right to left in the transform (high freqs -> low freqs), with blue ones skipped entirely.

Will the _orange_ filter produce maximum coefficients? Again, second order is just first order but over rows of `abs(CWT)` instead of over `x` (and less filters). If all goes according to plan, we'll see the maximum near the _bottom_ of second-order transform (low frequency), due to brown (left-most) filter. And indeed:

![image](https://user-images.githubusercontent.com/16495490/113699336-7c7ad000-96e6-11eb-9dff-4dd531badf88.png)

(**Important distinction**: this plot is _not_ interpretable same way was first-order, it's actually a stack of independent 2D features, where each 'patch' corresponds to convolution of a second-order filter with every row of first-order transform.)

If we're correct yet further, vertical slices of this 2D plot should look like Gaussians; all 2048 superimposed (and left-trimmed):

![image](https://user-images.githubusercontent.com/16495490/113699801-1b9fc780-96e7-11eb-9083-5a340bbe5a1f.png)

Close enough; the imperfections are due to lack of filter redundancy (i.e. non-continuous spacing of center frequencies).

### 3. Finally, frequential scattering

We wrap up by convolving along _frequency_ - namely, along frequency axis of every second-order filter's own second-order transform (see "important distinction"). Note that we only have non-negligible energy in the very last wavelet's transform (confirm this in the gifs), so we skip straight to it. Below is the exact second-order slice we'll convolve with, with proper subsampling (but before padding):

![image](https://user-images.githubusercontent.com/16495490/113700778-4d655e00-96e8-11eb-9ba7-0af3757ec8b6.png)

We convolve along each of the 128 time units, and each of those time units slices is approximately Gaussian. So maximum coefficients will yield from wavelets that have narrow time-domain support, i.e. wide freq-domain support) Reminder of frequential scattering filterbank (and here blues aren't skipped):

![image](https://user-images.githubusercontent.com/16495490/113701237-f318cd00-96e8-11eb-81d7-3b80a241ba45.png)

Let's see how this works out exactly; first we have padding and DFT along frequency:

![image](https://user-images.githubusercontent.com/16495490/113701676-80f4b800-96e9-11eb-9557-d1ccdb51459e.png)

Notice while the DFT is concentrated toward lower frequencies, it also has wide support (heatmap extends vertically), while wavelets closer to lower frequencies have narrower support. As we'll see, support plays greater role:

![image](https://user-images.githubusercontent.com/16495490/113705058-05e1d080-96ee-11eb-9722-bebd6daaf7cd.png)

(y-axis disparity per differing subsamplings; note the 1D wavelets multiply the images _vertically_) After lowpass + subsample, using color norm `(0, 7e-7)` as in the gif:

![image](https://user-images.githubusercontent.com/16495490/113707541-1b0c2e80-96f1-11eb-9375-a1b77929c51d.png)

Compare against the gif. Note the gif has redder Psi_7_ wavelets, meaning a filter between the two extremes correlates greater.

### 4. What do the coefficients tell us?

Mind above only covers one of three "pairs": `psi_t * psi_f`, where other two are `psi_t * phi_f` and `phi_t * psi_f`, where `phi` is a lowpass (and all pairs also get 2D lowpassed). In this example, it's actually more intuitive to inspect the other two pairs:

 1. `psi_t * phi_f`: last second order wavelet and lowpass frequential (left img below). The heatmap is concentrated at _bottom center_; why? Recall what frequential scattering is here: convolving a lowpass over second-order transform. i.e., over the "Gaussian". This Gaussian strip is located toward _bottom_ of second-order coefficients, thus the correlation's greatest there. The overall translation is: _"The input's amplitude modulation envelope oscillates at low frequency"_.
 2. `phi_t * psi_f`: this first smooths the first-order transform along time, lowering the original envelope's frequency just slightly (since it's already very low), then all frequential logic applies. Since the _shape_ of each timestep slice is the same (~Gauss), the only contributing factor is the amplitude, which again is a sinusoid: we get higher coefficients toward edges since that's where the amplitude envelope peaks. Note coefficients could've been greater toward center if there were other non-zero rows.
    - Notice also, the filter that "kills" most information has captured the most _accurate_ information (greater values toward edges). This is because the A.M. is closer to d.c. bin than any `psi2`'s center frequency, thus the lowpass captures it best.

![image](https://user-images.githubusercontent.com/16495490/113705574-a59f5e80-96ee-11eb-9ac7-3ecc971d0fb3.png)

### 6. Are up & down spins always same?

No; it is here per symmetry along vertical (frequential) axis (the "Gaussian") - but inspecting closely, they're not _exactly_ same, per imperfections in the Gaussian.

### 7. Notation

 - "Psi_1_2_+1" means `mu=1` (index of second-order wavelet), `l=2` (index of frequential wavelet), `spin = +1`. 
 - "Psi_-infinity_0..." means second-order wavelet is lowpass, i.e. `phi_t * psi_f` pair; spin is zero
 - "Psi_mu_-infinity_0..." means frequential wavelet is lowpass, i.e. `psi_t * phi_f` pair; spin is zero

### 8. Anything else?

Plenty; this is mainly a low-level internals overview. But it takes long so I wrap up here.

### 9. Going deeper

Optional reading

   - Natural question arises on the `psi_t * phi_f` image: why does it "taper off" toward sides (horizontally)? If it's Gaussian for every time unit, each row should be same color. The reason is, the Gaussians have different peak amplitudes across timeshifts, which horizontally trace out a rise-fall curve. Convolving with frequential lowpass yields a peak to left and to right of center, which, when lowpassed through time, averages to one peak in the center. _Why_ Gaussians' peaks differ is separate: spectral leakage + shadow frequencies combining into amplitude-modulated waveforms.

## To devs

If you run my code, you'll find some of the coefficients differing significantly from ones in the gifs. A quiz for devs, which also sets ground for a feature request: _why_?

With only one change in code, unrelated to anything stemming from `kymatio`, you'll get gifs' coefficients. The key coefficient difference is also shown.

<details>
  <summary><b>hint</b></summary>

I sampled `x` a little differently

</details>

<details>
  <summary><b>hint 2</b></summary>

The image shows my `x`

</details>

### Key difference

<img src="https://user-images.githubusercontent.com/16495490/113710740-25302c00-96f5-11eb-87a6-fcc89eba3120.png" width="600">

<details>
  <summary>code</summary>

```python
import numpy as np
import matplotlib.pyplot as plt
from kymatio.numpy import TimeFrequencyScattering
from ssqueezepy.visuals import imshow, plot

#%%
def gif_jtfs(Scx, meta, norms=(1e-4, 1e-6, 1e-4), inf_token=-1, skip_spins=False):
    norms = [(0, n) for n in norms]
    n_spins = sum(s for s in meta['s'][1:] if s == 1)

    i = 1  # skip first-order

    # `psi_t * psi_f` first
    while inf_token not in meta['n'][i + n_spins]:
        if not skip_spins:
            _viz_spins(Scx, meta, i, norms[0], n_spins)
        i += 1
    i = (i - 1) * 2 + 1

    # `psi_t * phi_f`
    while meta['n'][i][1] == inf_token:
        _viz_simple(Scx, meta, i, norms[1])
        i += 1

    # `phi_t * psi_f`
    while meta['n'][i][0] == inf_token:
        _viz_simple(Scx, meta, i, norms[2])
        i += 1
        if i >= len(Scx):
            break


def _title(meta, i, spin):
    txt = r"$|\Psi_{%s, %s, %s} \star X| \star \phi_T \star \phi_F$"
    mu, l = [int(n) if (float(n).is_integer() and n >= 0) else '\infty'
             for n in meta['n'][i]]

    return (txt % (mu, l, spin), {'fontsize': 20})

def _viz_spins(Scx, meta, i, norm, n_spins):
    i0, i1 = i, i + n_spins

    fig, axes = plt.subplots(1, 2, figsize=(15, 7))
    kw = dict(abs=1, ticks=0, show=0, norm=norm)
    imshow(Scx[i0]['coef'], ax=axes[0], **kw, title=_title(meta, i, '+1'))
    imshow(Scx[i1]['coef'], ax=axes[1], **kw, title=_title(meta, i, '-1'))
    plt.subplots_adjust(wspace=0.01)
    plt.show()

def _viz_simple(Scx, meta, i, norm):
    imshow(Scx[i]['coef'], abs=1, ticks=0, show=1, norm=norm, w=.8, h=.5,
           title=_title(meta, i, '0'))

#%%###########################################################################
T = 2048
J = 7
Q = 16

scattering = TimeFrequencyScattering(J, T, Q, J_fr=4, Q_fr=2, out_type='list')

#%%
f1, f2 = 1, 64
t = np.linspace(0, 1, T, 0)
a = (np.cos(2*np.pi * f1 * t) + 1) / 2
c = np.cos(2*np.pi * f2 * t)
x = a * c
plot(x, title="x | T=%s" % T)

Scx = scattering(x)

#%%
gif_jtfs(Scx, scattering.meta(),
         norms=(2e-7, 1e-7, 2e-5))
```
</details>

**Comment by lostanlen** [comment]
Very cool. Not sure what is happening towards the end but happy to consider your feature request
It does seem like something is amiss near the edges in the implementation on the right
